File: core/src/main/java/com/core/loitp/gallery/album/GalleryCoreAlbumActivity.java
package com.core.loitp.gallery.album;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.animation.OvershootInterpolator;
import android.widget.LinearLayout;

import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.core.base.BaseFontActivity;
import com.core.common.Constants;
import com.core.loitp.gallery.photos.GalleryCorePhotosActivity;
import com.core.utilities.LActivityUtil;
import com.core.utilities.LLog;
import com.core.utilities.LUIUtil;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import loitp.core.R;
import vn.loitp.restapi.flickr.FlickrConst;
import vn.loitp.restapi.flickr.model.photosetgetlist.Photoset;
import vn.loitp.restapi.flickr.service.FlickrService;
import vn.loitp.restapi.restclient.RestClient;
import vn.loitp.views.layout.floatdraglayout.DisplayUtil;
import vn.loitp.views.progressloadingview.avloadingindicatorview.AVLoadingIndicatorView;
import vn.loitp.views.recyclerview.animator.adapters.ScaleInAnimationAdapter;
import vn.loitp.views.recyclerview.animator.animators.SlideInRightAnimator;

public class GalleryCoreAlbumActivity extends BaseFontActivity {
    private AlbumAdapter albumAdapter;
    private List<Photoset> photosetList = new ArrayList<>();
    private ArrayList<String> removeAlbumList;
    private int bkgRootView;
    private AVLoadingIndicatorView avLoadingIndicatorView;
    private AdView adView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        isShowAdWhenExit = false;
        setTransparentStatusNavigationBar();
        removeAlbumList = getIntent().getStringArrayListExtra(Constants.getKEY_REMOVE_ALBUM_FLICKR_LIST());

        avLoadingIndicatorView = findViewById(R.id.av);

        final String admobBannerUnitId = getIntent().getStringExtra(Constants.getAD_UNIT_ID_BANNER());
        LLog.d(TAG, "admobBannerUnitId " + admobBannerUnitId);
        final LinearLayout lnAdview = findViewById(R.id.ln_adview);
        if (admobBannerUnitId == null || admobBannerUnitId.isEmpty()) {
            lnAdview.setVisibility(View.GONE);
        } else {
            adView = new AdView(activity);
            adView.setAdSize(AdSize.BANNER);
            adView.setAdUnitId(admobBannerUnitId);
            LUIUtil.createAdBanner(adView);
            lnAdview.addView(adView);
            int navigationHeight = DisplayUtil.getNavigationBarHeight(activity);
            LUIUtil.setMargins(lnAdview, 0, 0, 0, navigationHeight + navigationHeight / 4);
        }

        final RecyclerView recyclerView = findViewById(R.id.recyclerView);

        bkgRootView = getIntent().getIntExtra(Constants.getBKG_ROOT_VIEW(), Constants.getNOT_FOUND());
        LLog.d(TAG, "bkgRootView " + bkgRootView);
        if (bkgRootView == Constants.getNOT_FOUND()) {
            getRootView().setBackgroundColor(ContextCompat.getColor(activity, R.color.colorPrimary));
        } else {
            getRootView().setBackgroundResource(bkgRootView);
        }

        final SlideInRightAnimator animator = new SlideInRightAnimator(new OvershootInterpolator(1f));
        animator.setAddDuration(1000);
        recyclerView.setItemAnimator(animator);

        recyclerView.setLayoutManager(new LinearLayoutManager(activity));
        recyclerView.setHasFixedSize(true);

        albumAdapter = new AlbumAdapter(activity, photosetList, new AlbumAdapter.Callback() {
            @Override
            public void onClick(int pos) {
                final Intent intent = new Intent(activity, GalleryCorePhotosActivity.class);
                intent.putExtra(Constants.getBKG_ROOT_VIEW(), bkgRootView);
                intent.putExtra(Constants.getSK_PHOTOSET_ID(), photosetList.get(pos).getId());
                intent.putExtra(Constants.getSK_PHOTOSET_SIZE(), photosetList.get(pos).getPhotos());
                startActivity(intent);
                LActivityUtil.tranIn(activity);
            }

            @Override
            public void onLongClick(int pos) {
                /*photosetList.remove(pos);
                albumAdapter.notifyItemRemoved(pos);
                albumAdapter.notifyItemRangeChanged(pos, photosetList.size());*/
            }
        });
        //recyclerView.setAdapter(albumAdapter);
        final ScaleInAnimationAdapter scaleAdapter = new ScaleInAnimationAdapter(albumAdapter);
        scaleAdapter.setDuration(1000);
        scaleAdapter.setInterpolator(new OvershootInterpolator());
        scaleAdapter.setFirstOnly(true);
        recyclerView.setAdapter(scaleAdapter);

        LUIUtil.setPullLikeIOSVertical(recyclerView);

        photosetsGetList();
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return "TAG" + getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_gallery_core_album;
    }

    private void photosetsGetList() {
        avLoadingIndicatorView.smoothToShow();
        final FlickrService service = RestClient.createService(FlickrService.class);
        final String method = FlickrConst.METHOD_PHOTOSETS_GETLIST;
        final String apiKey = FlickrConst.API_KEY;
        final String userID = FlickrConst.USER_KEY;
        final int page = 1;
        final int perPage = 500;
        final String primaryPhotoExtras = FlickrConst.PRIMARY_PHOTO_EXTRAS_0;
        final String format = FlickrConst.FORMAT;
        final int nojsoncallback = FlickrConst.NO_JSON_CALLBACK;


        compositeDisposable.add(service.photosetsGetList(method, apiKey, userID, page, perPage, primaryPhotoExtras, format, nojsoncallback)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(wrapperPhotosetGetlist -> {
                    //LLog.d(TAG, "onSuccess " + new Gson().toJson(wrapperPhotosetGetlist));
                    photosetList.addAll(wrapperPhotosetGetlist.getPhotosets().getPhotoset());

                    /*String x = "";
                    for (int i = 0; i < photosetList.size(); i++) {
                        x += photosetList.get(i).getTitle().getContent() + " - " + photosetList.get(i).getId() + "\n";
                    }
                    LLog.d(TAG, "" + x);*/

                    //LLog.d(TAG, "orginal size: " + photosetList.size());
                    //LLog.d(TAG, "removeAlbumList size: " + removeAlbumList.size());
                    for (int i = removeAlbumList.size() - 1; i >= 0; i--) {
                        for (int j = photosetList.size() - 1; j >= 0; j--) {
                            if (removeAlbumList.get(i).equals(photosetList.get(j).getId())) {
                                photosetList.remove(j);
                            }
                        }
                    }
                    //LLog.d(TAG, "after size: " + photosetList.size());
                    Collections.sort(photosetList, new Comparator<Photoset>() {
                        @Override
                        public int compare(Photoset o1, Photoset o2) {
                            return Long.valueOf(o2.getDateUpdate()).compareTo(Long.valueOf(o1.getDateUpdate()));
                        }
                    });
                    updateAllViews();
                    avLoadingIndicatorView.smoothToHide();
                }, throwable -> {
                    handleException(throwable);
                    avLoadingIndicatorView.smoothToHide();
                }));
    }

    private void updateAllViews() {
        if (albumAdapter != null) {
            albumAdapter.notifyDataSetChanged();
        }
    }

    @Override
    protected void onResume() {
        if (adView != null) {
            adView.resume();
        }
        super.onResume();
    }

    @Override
    public void onPause() {
        if (adView != null) {
            adView.pause();
        }
        super.onPause();
    }

    @Override
    public void onDestroy() {
        if (adView != null) {
            adView.destroy();
        }
        super.onDestroy();
    }
}

File: core/src/main/java/com/core/loitp/gallery/album/GalleryCoreAlbumActivity.kt
package com.core.loitp.gallery.album

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.animation.OvershootInterpolator
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.core.base.BaseFontActivity
import com.core.common.Constants
import com.core.loitp.gallery.photos.GalleryCorePhotosActivity
import com.core.utilities.LActivityUtil
import com.core.utilities.LLog
import com.core.utilities.LUIUtil
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.restapi.flickr.FlickrConst
import com.restapi.flickr.model.photosetgetlist.Photoset
import com.restapi.flickr.service.FlickrService
import com.restapi.restclient.RestClient
import com.views.layout.floatdraglayout.DisplayUtil
import com.views.progressloadingview.avloadingindicatorview.AVLoadingIndicatorView
import com.views.recyclerview.animator.adapters.ScaleInAnimationAdapter
import com.views.recyclerview.animator.animators.SlideInRightAnimator
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import loitp.core.R
import java.util.*

class GalleryCoreAlbumActivity : BaseFontActivity() {
    private var albumAdapter: AlbumAdapter? = null
    private val photosetList = ArrayList<Photoset>()
    private var removeAlbumList: ArrayList<String>? = null
    private var bkgRootView: Int = 0
    private var avLoadingIndicatorView: AVLoadingIndicatorView? = null
    private var adView: AdView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isShowAdWhenExit = false
        setTransparentStatusNavigationBar()
        removeAlbumList = intent.getStringArrayListExtra(Constants.KEY_REMOVE_ALBUM_FLICKR_LIST)
        avLoadingIndicatorView = findViewById(R.id.av)
        val admobBannerUnitId = intent.getStringExtra(Constants.AD_UNIT_ID_BANNER)
        LLog.d(TAG, "admobBannerUnitId $admobBannerUnitId")
        val lnAdview = findViewById<LinearLayout>(R.id.ln_adview)
        if (admobBannerUnitId.isNullOrEmpty()) {
            lnAdview.visibility = View.GONE
        } else {
            adView = AdView(activity)
            adView?.let {
                it.adSize = AdSize.BANNER
                it.adUnitId = admobBannerUnitId
                LUIUtil.createAdBanner(it)
                lnAdview.addView(it)
                val navigationHeight = DisplayUtil.getNavigationBarHeight(activity)
                LUIUtil.setMargins(lnAdview, 0, 0, 0, navigationHeight + navigationHeight / 4)
            }
        }

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        bkgRootView = intent.getIntExtra(Constants.BKG_ROOT_VIEW, Constants.NOT_FOUND)
        LLog.d(TAG, "bkgRootView $bkgRootView")
        if (bkgRootView == Constants.NOT_FOUND) {
            rootView?.setBackgroundColor(ContextCompat.getColor(activity, R.color.colorPrimary))
        } else {
            rootView?.setBackgroundResource(bkgRootView)
        }

        val animator = SlideInRightAnimator(OvershootInterpolator(1f))
        animator.addDuration = 1000

        recyclerView?.let {
            it.itemAnimator = animator
            it.layoutManager = LinearLayoutManager(activity)
            it.setHasFixedSize(true)
        }

        albumAdapter = AlbumAdapter(activity, photosetList, object : AlbumAdapter.Callback {
            override fun onClick(pos: Int) {
                val intent = Intent(activity, GalleryCorePhotosActivity::class.java)
                intent.apply {
                    putExtra(Constants.BKG_ROOT_VIEW, bkgRootView)
                    putExtra(Constants.SK_PHOTOSET_ID, photosetList[pos].id)
                    putExtra(Constants.SK_PHOTOSET_SIZE, photosetList[pos].photos)
                    startActivity(this)
                    LActivityUtil.tranIn(activity)
                }
            }

            override fun onLongClick(pos: Int) {
                /*photosetList.remove(pos);
                albumAdapter.notifyItemRemoved(pos);
                albumAdapter.notifyItemRangeChanged(pos, photosetList.size());*/
            }
        })
        val scaleAdapter = ScaleInAnimationAdapter(albumAdapter)
        scaleAdapter.apply {
            setDuration(1000)
            setInterpolator(OvershootInterpolator())
            setFirstOnly(true)
        }

        recyclerView?.let {
            it.adapter = scaleAdapter
            LUIUtil.setPullLikeIOSVertical(it)
        }

        photosetsGetList()
    }

    override fun setFullScreen(): Boolean {
        return false
    }

    override fun setTag(): String? {
        return "TAG" + javaClass.simpleName
    }

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_gallery_core_album
    }

    private fun photosetsGetList() {
        avLoadingIndicatorView?.smoothToShow()
        val service = RestClient.createService(FlickrService::class.java)
        val method = FlickrConst.METHOD_PHOTOSETS_GETLIST
        val apiKey = FlickrConst.API_KEY
        val userID = FlickrConst.USER_KEY
        val page = 1
        val perPage = 500
        val primaryPhotoExtras = FlickrConst.PRIMARY_PHOTO_EXTRAS_0
        val format = FlickrConst.FORMAT
        val nojsoncallback = FlickrConst.NO_JSON_CALLBACK

        compositeDisposable.add(service.photosetsGetList(method, apiKey, userID, page, perPage, primaryPhotoExtras,
                format, nojsoncallback)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ wrapperPhotosetGetlist ->
                    //LLog.d(TAG, "onSuccess " + new Gson().toJson(wrapperPhotosetGetlist));
                    photosetList.addAll(wrapperPhotosetGetlist.photosets.photoset)

                    /*String x = "";
                    for (int i = 0; i < photosetList.size(); i++) {
                        x += photosetList.get(i).getTitle().getContent() + " - " + photosetList.get(i).getId() + "\n";
                    }
                    LLog.d(TAG, "" + x);*/

                    //LLog.d(TAG, "orginal size: " + photosetList.size());
                    //LLog.d(TAG, "removeAlbumList size: " + removeAlbumList.size());

                    removeAlbumList?.let {
                        for (i in it.indices.reversed()) {
                            for (j in photosetList.indices.reversed()) {
                                if (it[i] == photosetList[j].id) {
                                    photosetList.removeAt(j)
                                }
                            }
                        }
                    }

                    //LLog.d(TAG, "after size: " + photosetList.size());
                    photosetList.sortWith(Comparator { o1, o2 ->
                        java.lang.Long.valueOf(o2.dateUpdate).compareTo(java.lang.Long.valueOf(o1.dateUpdate))
                    })
                    updateAllViews()
                    avLoadingIndicatorView?.smoothToHide()
                }, { throwable ->
                    handleException(throwable)
                    avLoadingIndicatorView?.smoothToHide()
                }))
    }

    private fun updateAllViews() {
        albumAdapter?.notifyDataSetChanged()
    }

    override fun onResume() {
        adView?.resume()
        super.onResume()
    }

    public override fun onPause() {
        adView?.pause()
        super.onPause()
    }

    public override fun onDestroy() {
        adView?.destroy()
        super.onDestroy()
    }
}
